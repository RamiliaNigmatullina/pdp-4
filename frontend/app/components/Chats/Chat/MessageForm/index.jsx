import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import $ from 'jquery';
import 'form-serializer';
import { createMessage } from '../../api/index';
import styles from './styles.module.scss';

class MessageForm extends React.Component {
  constructor(props) {
    super(props);
    this.textInput = React.createRef();

    this.state = { isSending: false, value: this.getValue() };
  }

  componentDidUpdate(prevProps) {
    if (prevProps.chatId !== this.props.chatId) {
      this.setState({ value: this.getValue() });
    }
  }

  renderTextarea() {
    return (
      <div className={classNames(styles.formText, 'form-group', 'text')}>
        <textarea
          className="form-control"
          name="message[text]"
          onChange={(event) => { this.handleChange(event); }}
          onKeyUp={this.handleKeyUp}
          placeholder="Type text..."
          ref={this.textInput}
          rows="1"
          value={this.state.value}
        />
      </div>
    );
  }

  render() {
    return (
      <form
        className={classNames(styles.messageForm, 'simple_form')}
        method="post"
        action={`/api/v1/chats/${this.props.chatId}/messages`}
        onSubmit={this.handleSubmit}
      >

        { this.renderTextarea() }
        <button
          className={classNames(styles.formButton, 'btn', 'btn-primary')}
          disabled={this.state.isSending}>
          Send
        </button>
      </form>
    );
  }

  handleSubmit = (event) => {
    event.preventDefault();

    if (!this.textInput.current.value.trim() || this.state.isSending) return;

    this.setState({ isSending: true, value: this.textInput.current.value.trim() });

    const formData = $(event.target).serializeObject();
    createMessage(this.props.chatId, formData)
      .then(() => {
        localStorage.setItem(this.chatUid(), '');
        this.setState({ isSending: false, value: '' });
      });
  }

  handleKeyUp = (event) => {
    if (event.keyCode === 13) {
      this.setState({ value: this.textInput.current.value.trim() });

      this.handleSubmit(event);
    }
  }

  handleChange = (event) => {
    const { value } = event.target;

    localStorage.setItem(this.chatUid(), value);
    this.setState({ value });
  }

  chatUid = () => {
    const { chatId, currentUser } = this.props;

    return `chat_${chatId}_${currentUser.id}`;
  }

  getValue = () => localStorage.getItem(this.chatUid()) || ''
}

MessageForm.propTypes = {
  chatId: PropTypes.number.isRequired,
  currentUser: PropTypes.object.isRequired,
};

export default MessageForm;
