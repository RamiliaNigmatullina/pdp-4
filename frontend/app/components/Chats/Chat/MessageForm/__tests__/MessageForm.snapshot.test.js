import React from 'react';
import { shallow } from 'enzyme';
import MessageForm from '../index';

describe('MessageForm', () => {
  let props;

  const render = () => shallow(<MessageForm {...props} />);

  beforeEach(() => {
    props = {
      chatId: 1,
      currentUser: {
        id: 1,
        full_name: 'John Smith',
        avatar: '/john_smith_avatar.jpg',
      },
    };
  });

  it('renders message form', () => {
    const component = render();

    expect(component).toMatchSnapshot();
  });
});
