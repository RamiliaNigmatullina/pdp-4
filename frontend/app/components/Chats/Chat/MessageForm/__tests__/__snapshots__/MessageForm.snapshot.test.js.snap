// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageForm renders message form 1`] = `
<form
  action="/api/v1/chats/1/messages"
  className="messageForm simple_form"
  method="post"
  onSubmit={[Function]}
>
  <div
    className="formText form-group text"
  >
    <textarea
      className="form-control"
      name="message[text]"
      onChange={[Function]}
      onKeyUp={[Function]}
      placeholder="Type text..."
      rows="1"
      value=""
    />
  </div>
  <button
    className="formButton btn btn-primary"
    disabled={false}
  >
    Send
  </button>
</form>
`;
